### PR TITLE
Strict null on some query and connection

### PR DESCRIPTION
**Pull request by anthonydresser**

### CODE REVIEW

File: src/sql/platform/connection/common/connectionConfig.ts [hunk 1]
```diff
@@ -348,21 +354,21 @@ export class ConnectionConfig {
...
 		if (!groupTree) {
 			groupTree = [];
 		}
-		let newGroupId: string;
+		let newGroupId: string | undefined;
 
 		if (index < groupNames.length) {
 			let groupName: string = groupNames[index];
-			let newGroup: IConnectionProfileGroup = {
+			let newGroup = <unknown>{ // janky workaround to make this work properly
```
Comment (Charles-Gagnon) [inline src/sql/platform/connection/common/connectionConfig.ts]: >janky [](start = 32, length = 5)

nit: but let's avoid using slang like this - Policheck tends to not like stuff like that

File: src/sql/platform/connection/common/connectionProfile.ts [hunk 2]
```diff
@@ -215,14 +215,11 @@ export class ConnectionProfile extends ProviderConnectionInfo implements interfa
 	}
 
 	public static fromIConnectionProfile(capabilitiesService: ICapabilitiesService, profile: azdata.IConnectionProfile) {
```
Comment (Charles-Gagnon) [inline src/sql/platform/connection/common/connectionProfile.ts]: >fromIConnectionProfile [](start = 15, length = 22)

Could you add the return type to this while you're here?

File: src/sql/platform/connection/common/connectionProfile.ts [hunk 3]
```diff
@@ -215,14 +215,11 @@ export class ConnectionProfile extends ProviderConnectionInfo implements interfa
...
-			if (profile instanceof ConnectionProfile) {
-				return profile;
-			} else {
-				return new ConnectionProfile(capabilitiesService, profile);
-			}
+		if (profile instanceof ConnectionProfile) {
+			return profile;
+		} else {
+			return new ConnectionProfile(capabilitiesService, profile);
 		}
```
Comment (Charles-Gagnon) [inline src/sql/platform/connection/common/connectionProfile.ts]: You're changing the behavior here - is that intended? I don't really see a huge problem with this change but let's make sure nothing is depending on this functionality

Comment (anthonydresser) [reply src/sql/platform/connection/common/connectionProfile.ts]: Yes, it simplifies the actual change I have to make to have it meet strict null. I checked the callers of this function (and others like this) and verified they aren't just randomly passing in undefined.

File: src/sql/platform/query/common/queryRunner.ts [hunk 4]
```diff
@@ -135,39 +135,43 @@ export default class QueryRunner extends Disposable {
...
 	 */
-	public runQuery(input: azdata.ISelectionData, runOptions?: azdata.ExecutionPlanOptions): Thenable<void>;
-	public runQuery(input, runOptions?: azdata.ExecutionPlanOptions): Thenable<void> {
-		return this.doRunQuery(input, false, runOptions);
+	public runQuery(input: azdata.ISelectionData, runOptions?: azdata.ExecutionPlanOptions): Promise<void>;
+	public runQuery(input: string | azdata.ISelectionData, runOptions?: azdata.ExecutionPlanOptions): Promise<void> {
+		if (types.isString(input)) {
+			return this.doRunQuery(input, false, runOptions);
+		} else {
+			return this.doRunQuery(input, false, runOptions);
```
Comment (Charles-Gagnon) [inline src/sql/platform/query/common/queryRunner.ts]: Interesting - why is this necessary? Shouldn't it be able to use the doRunQuery that takes in a string | azdata.ISelectionData directly?

**Comment by Charles-Gagnon** [comment]
> 			return Promise.reject('Unknown input');

This function returns a Promise<void> so we shouldn't be resolving (above) or rejecting with values like this. 

---
Refers to: src/sql/platform/query/common/queryRunner.ts:208 in a4d5c09. [](commit_id = a4d5c09e04cd2ab42b24d77f73be5554398b5bf1, deletion_comment = False)

File: src/sql/platform/query/common/queryRunner.ts (hunk 4 above)
Comment (anthonydresser) [reply src/sql/platform/query/common/queryRunner.ts]: Its a weird behavior of the typescript compiler, even though the inputs are the same in terms of overloading it doesn't resolve it as being correct, its basically an unnecessary check to make the typescript compiler happy. It's definitely a bug in the compiler (although there are a lot of weird behavior like this that actually has some reasoning).

File: src/sql/workbench/parts/query/common/resultSerializer.ts [hunk 5]
```diff
@@ -212,7 +212,7 @@ export class ResultSerializer {
 		} else if (format === SaveFormat.XML) {
 			saveResultsParams = this.getConfigForXml();
 		}
-		return saveResultsParams;
+		return saveResultsParams!; // this could be unsafe
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/query/common/resultSerializer.ts]: >// this could be unsafe [](start = 29, length = 23)

Feels like we should log an error at the very least or something if it's an unknown save format

**Comment by anthonydresser** [comment]
I'm not sure what you mean by we shouldn't be resolving or rejecting with values?

**Comment by Charles-Gagnon** [comment]
Ugh, dunno why CodeFlow put that as a general comment...anyways for the resolving/rejecting the return type for the function is a Promise<void> but on line 176 we're doing

`return Promise.resolve(undefined);` 

which we've discussed - but I consider to be a very different thing than void and so resolving that there is just odd.

Nevermind about the reject, that was me just misunderstanding.
